classpath/avian/http/Handler.java

```diff
@@ -10,17 +10,112 @@
 
 package avian.http;
 
-import java.net.URL;
-import java.net.URLStreamHandler;
-import java.net.URLConnection;
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.ByteArrayInputStream;
 import java.io.IOException;
-import java.io.FileNotFoundException;
-import java.io.File;
-import java.io.FileInputStream;
 import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.net.Socket;
+import java.net.URL;
+import java.net.URLConnection;
+import java.net.URLStreamHandler;
+import java.util.HashMap;
+import java.util.Map;
+
+public class Handler extends URLStreamHandler
+{
+    public URLConnection openConnection(URL url) throws IOException
+    {
+        return new HttpURLConnection(url);
+    }
+    
+    class HttpURLConnection extends URLConnection
+    {
+        Socket socket;
+        private BufferedWriter writer;
+        private InputStream bin;
+        private Map<String,String> header = new HashMap<String, String>();
+        private int status;
+        
+        protected HttpURLConnection(URL url)
+        {
+            super(url);
+        }
+
+        @Override
+        public void connect() throws IOException
+        {
+            if(socket == null)
+            {
+                URLConnection con = null;
+                String host = url.getHost();
+                int port =url.getPort();
+                if(port < 0) port = 80;
+                socket = new Socket(host, port);
+                OutputStream out = socket.getOutputStream();
+                writer = new BufferedWriter(new OutputStreamWriter(out));
+                writer.write("GET " + url.getPath() + " HTTP/1.1");
+                writer.write("\nHost: " + host);
+                writer.write("\n\n");
+                writer.flush();
+                bin = new BufferedInputStream(socket.getInputStream());
+                readHeader();
+//                System.out.println("Status: " + status);
+//                System.out.println("Headers: " + header);
+            }
+        }
+
+        private void readHeader() throws IOException
```

Comment:
With regard to point 3, once upon a time, when implementing an internet standard, the prevailing wisdom was to be liberal in what you accept and strict/conservative in what you produce.  That backfired because people NOT following the spec (i.e. just trying to get things working) introduced subtle non-compliance into their code, simply because it worked with the (liberal) endpoint they happened to be using at the time.  Thus, the internet is littered with all sorts of bizarre almost-correct implementations of various protocols.  IMO, that principle turned out to be bad for the long-term sanity of the internet.

Given that, there are two opposing arguments to be made here:

* That we ought to implement something that doesn't take any bullshit, in an effort to put a little pressure on other people implementing the protocol to make sure it's correct.
* That we ought to cave, and implement something like what you have here (i.e. be liberal in what we accept).  If we went down this path, I would say we should be allowing either '\r\n\r\n' (correct) and '\n\n' (a common mistake) as the terminator.

I would opt for the former, as long as we don't immediately see any compatibility problems.

However, that argument is largely academic at this point.  I just wanted to give a little background.

The much more important problem to fix is the byte-by-byte reading, as I pointed out below.